[serve][docs] Add dev workflow page (#27746)
ray-project/ray: doc/source/serve/dev-workflow.md

Now that Serve is running, we can send HTTP requests to the application.
For simplicity, we'll just use the `curl` command to send requests from another terminal.

```bash
curl -X PUT "http://localhost:8000/?name=Ray"
# Hello, Ray! Hello, Ray!
```

After you're done testing, you can shut down Ray Serve by interrupting the `serve run` command (e.g., with Ctrl-C):

```console
^C2022-08-11 11:47:19,829       INFO scripts.py:323 -- Got KeyboardInterrupt, shutting down...
(ServeController pid=9865) INFO 2022-08-11 11:47:19,926 controller 9865 deployment_state.py:1257 - Removing 1 replicas from deployment 'Doubler'.
(ServeController pid=9865) INFO 2022-08-11 11:47:19,929 controller 9865 deployment_state.py:1257 - Removing 1 replicas from deployment 'HelloDeployment'.
```

## Testing on a remote cluster

To test on a remote cluster, you'll use `serve run` again, but this time you'll pass in an `--address` argument to specify the address of the Ray cluster to connect to.  For remote clusters, this address has the form `ray://<head-node-ip-address>:10001`; see [Ray Client](ray-client-under-construction) for more information.

> When making the transition from your local machine to a remote cluster, you'll need to make sure your cluster has a similar environment to your local machine--files, environment variables, and Python packages, for example.  During development, you can use {ref}`Runtime Environments<runtime-environments>` to manage these dependencies.
> 
> Let's see a simple example. Run the following command on your local machine, with your remote cluster head node IP address substituted for `<head-node-ip-address>` in the command:
> 
> ```bash
> serve run  --address=ray://<head-node-ip-address>:10001 --working_dir="./project/src" local_dev:HelloDeployment
> ```
> 
> This will connect to the remote cluster via Ray Client, upload the `working_dir` directory, and run your serve application.  Here, the local directory specified by `working_dir` must contain `local_dev.py` so that it can be uploaded to the cluster and imported by Ray Serve.
> 
> Once this is up and running, we can send requests to the application:
> 
> ```bash
> curl -X PUT http://<head-node-ip-address>:8000/?name=Ray
> # Hello, Ray! Hello, Ray!
> ```
> 
> If your runtime environment contains fields other than `working_dir`, use the `--runtime-env-json` argument:

Reviewer: ```suggestion
When making the transition from your local machine to a remote cluster, you'll need to make sure your cluster has a similar environment to your local machine--files, environment variables, and Python packages, for example.

Let's see a simple example that just packages the code. Run the following command on your local machine, with your remote cluster head node IP address substituted for `<head-node-ip-address>` in the command:

```bash
serve run  --address=ray://<head-node-ip-address>:10001 --working_dir="<directory containing local_dev.py>" local_dev:HelloDeployment
```
This will connect to the remote cluster via Ray Client, upload the directory specified by `--working_dir`, and run your Serve application in this directory. We can make `local_dev.py` importable by Ray Serve by making sure that it is contained in the directory passed to `--working-dir`.

Once this is up and running, we can send requests to the application:

```bash
curl -X PUT http://<head-node-ip-address>:8000/?name=Ray
# Hello, Ray! Hello, Ray!
```

For more complex dependencies, including files outside the working directory, environment variables, and Python packages, you can use {ref}`Runtime Environments<runtime-environments>` to manage these dependencies. Here is an example using the `--runtime-env-json` argument:
```

Reviewer: Oh no, I think my suggestion got mis-formatted by GitHub :dagger: 

But basically I was trying to:
1. move the mention of runtime envs closer to where it's actually used in a command (for a first-time user, I don't think it will be obvious that --working-dir is just syntactic sugar for a runtime env)
2. explain semantics of `--working-dir`

Author: Makes sense!

> % TODO(architkulkarni): Add in the appropriate sidebars and table of contents

Author: No, I've removed it though

% TODO(architkulkarni): Add in the appropriate sidebars and table of contents

# Sample Development Workflow

> When developing a new application, iteration speed is critical.  This page describes a sample development workflow for Ray Serve applications, optimized for iteration speed. If you're ready to go to production, you can jump to the [Production Workflow](production-workflow.md) section.

Reviewer 2: ```suggestion
This page describes the recommended development workflow for Ray Serve applications when iterating in development. Once you're ready to go to production, you can jump to the [Production Guide](production-guide.md) section.
```

too much exposition for documentation here IMO. let's keep it to the point

## Project structure

There are a few reasonable options for organizing your project structure, depending on the complexity of your application.  

- You can put all of your code in a single Python file.

> - You can put your call graph in a Python file and your deployment class definitions in other files.

Reviewer 2: let's just drop this since it isn't any different than a typical python project

> ## Local Development using Python handles

Reviewer 2: ```suggestion
## Local Development using `serve.run`
```

You can use `serve.run` in process to run and test your application locally, using a Python handle to send requests rather than HTTP.

Benefits:

- Self-contained Python is convenient for writing local integration tests.
- No need to deploy to a cloud provider or manage infrastructure.

Drawbacks:

- Doesn't test HTTP endpoints.
- Doesn't test multi-node clusters, can't use GPUs if your local machine doesn't have them.

Let's see a simple example.

> ```python

Reviewer 3: move the code to the doc_code dir?

# Filename: local_dev.py
from ray import serve
from ray.dag.input_node import InputNode
import starlette.requests

@serve.deployment
def double(s: str):
    return s + " " + s

@serve.deployment
class HelloDeployment:
    def __call__(self, request: starlette.requests.Request):
        return "Hello, {}!".format(request.params["name"])

# Everything beyond this line could be in a different file, provided
# HelloDeployment and double are imported.
hello_deployment = HelloDeployment.bind()
with InputNode() as http_request:
    hello_output = hello_deployment.bind(http_request)

>     double_output = double.bind(hello_output)

Author: When I run this I get an error about calling `bind.()` twice on a ClassNode

Reviewer 4: I think this line:

```
hello_output = hello_deployment.bind(http_request)
```

should be this:

```
hello_output = hello_deployment.__call__.bind(http_request)
```

Could you try that and see if it fixes the issue?

ref = double_output.execute("Ray")
result = ray.get(ref)
assert result = "Hello, Ray! Hello, Ray!"
```

Here we used the `execute` method to directly send a Python object to the deployment, rather than an HTTP request.

> % TODO: We didn't use serve.run() like we said at the begining; should we?

Author: > actually, maybe we should just not use deployment graph at all here and make it a simpler two-deployment app?

I think this is the key question, or at least the main thing I'm confused about.  Are we saying that an expected part of the dev to prod transition is rewriting the code to go from `serve.run` to using a deployment graph?

How do you actually make a two-deployment pipeline app without using a deployment graph?  You'd need to pass in the second deployment in a constructor, right?  But to then later convert it to a graph, you'd need to rewrite all the deployment code to remove the deployments from the constructors and use `InputNode`, which seems like a lot of work.

> % TODO: We didn't use serve.run() like we said at the begining; should we?

Reviewer 3: Another perspective, will using `serve.run` in dev slow down the velocity or `execute` will bring other benefits ?  If not, I don't think we need to provide the `execute` information in the doc.

## Local Development with HTTP requests

> You can use the `serve run` CLI command to run and test your application locally using HTTP to send requests (similar to how you might use `uvicorn run` if you're familiar with Uvicorn):

Reviewer 4: Can you mention somewhere that this command (1) is blocking and (2) can be killed with `ctrl-c`?

```bash
serve run local_dev:HelloDeployment
```

Now that Serve is running, we can pass in HTTP requests to the application and see the output.  For simplicity, we'll just use the `curl` command to send requests from another terminal.

```bash
curl -X GET http://localhost:8000/hello?name=Ray
```

## 3. Testing on a remote cluster

When making the transition from your local machine to a remote cluster, you'll need to make sure your cluster has a similar environment to your local machine--files, environment variables, and Python packages, for example.  During development, you can use [Runtime Environments](runtime-environments.md) to manage this in a flexible way.

> See Ray Client (TODO: link) for more information on the Ray address specified here by the `--address` parameter.

Reviewer 4: Make sure to replace this `TODO` with a link.

```bash

> serve run  --address=ray://<cluster-ip-address>:10001 --runtime-env-json='{"env_vars": {"MY_ENV_VAR": "my-value"}, "working_dir": "./project/src", "pip": ["requests", "chess"]}' local_dev:HelloDeployment

Reviewer 3: can we add a `note` what ./project/src should usually include? Or Add a link for `working_dir`

```

Once this is up and running, we can send requests to the application and see the output.

```bash
curl -X GET http://<cluster-ip-address>:8000/hello?name=Ray
```

For more complex runtime environments, you can pass in a YAML file; see [serve run](serve_cli.md#serve-run) for details.

> :::{tip}

Reviewer 3: this `tip`  worth a simple example to explain a bit more. if we already have examples, better to add a link.

This page describes the recommended development workflow for Ray Serve applications when iterating in development. Once you're ready to go to production, you can jump to the [Production Guide](production.md) section.

## Local Development using `serve.run`

You can use `serve.run` in a Python script to run and test your application locally, using a handle to send requests rather than HTTP.

Benefits:

- Self-contained Python is convenient for writing local integration tests.
- No need to deploy to a cloud provider or manage infrastructure.

Drawbacks:

- Doesn't test HTTP endpoints.
- Can't use GPUs if your local machine doesn't have them.

Let's see a simple example.

```{literalinclude} ../serve/doc_code/local_dev.py
:start-after: __local_dev_start__
:end-before: __local_dev_end__
:language: python
```

We can add the testing code below:

```{literalinclude} ../serve/doc_code/local_dev.py
:start-after: __local_dev_handle_start__
:end-before: __local_dev_handle_end__
:language: python
```

This test code could be in the same file or a different file.  Before moving to the next section, remove or comment out the local testing code above if you added it in the same file.

> 

Author: Good point, removed

Now that Serve is running, we can pass in HTTP requests to the application and see the output.
For simplicity, we'll just use the `curl` command to send requests from another terminal.

```bash
curl -X PUT "http://localhost:8000/?name=Ray"   
```

This will output:

```console
Hello, Ray! Hello, Ray!
```

After you're done testing, you can shut down Ray Serve with Ctrl-C:

```console
^C2022-08-11 11:47:19,829       INFO scripts.py:323 -- Got KeyboardInterrupt, shutting down...
(ServeController pid=9865) INFO 2022-08-11 11:47:19,926 controller 9865 deployment_state.py:1257 - Removing 1 replicas from deployment 'Doubler'.
(ServeController pid=9865) INFO 2022-08-11 11:47:19,929 controller 9865 deployment_state.py:1257 - Removing 1 replicas from deployment 'HelloDeployment'.
```

## Testing on a remote cluster

When making the transition from your local machine to a remote cluster, you'll need to make sure your cluster has a similar environment to your local machine--files, environment variables, and Python packages, for example.  During development, you can use {ref}`Runtime Environments<runtime-environments>` to manage this in a flexible way.

> See [Ray Client](ray-client-under-construction) for more information on the Ray address specified here by the `--address` parameter.

Author: Under the hood, `--address` is passed to `ray.init()` on the laptop.  If you pass in a `ray://` address, it will use Ray Client.  You could also specify a local Ray address, but for simplicity I'll leave that out of the guide.

Let's see a simple example.

```bash

> serve run  --address=ray://<cluster-ip-address>:10001 --runtime-env-json='{"env_vars": {"MY_ENV_VAR": "my-value"}, "working_dir": "./project/src", "pip": ["requests", "chess"]}' local_dev:HelloDeployment

Reviewer: Can we split this example into multiple steps to make it more digestible? Concrete suggestion:

1. First explain how to run Serve on a remote cluster, no runtime env involved.
2. Then show the common case that's currently mentioned at the end: "A common pattern is to use the root directory of your project as the `working_dir` of your runtime environment when testing on a remote cluster."
3. End with the advanced case where a full runtime env json / yaml is used.

Author: For step 1, I don't think there's a dev workflow that doesn't use runtime_env, because we need runtime_env to get the local code onto the cluster.  

Why don't I try this ordering to make it more digestible:
1. Just use `--working-dir`
2. Use a full `--runtime-env-json`
3. Mention runtime_env YAML

Reviewer: Ah I see. Yes, that flow sounds good, thanks!

Author: I gave this a shot, let me know how it looks

```

This will connect to the remote cluster via Ray Client and run your serve application.  Here, the directory specified by `working_dir` should contain `local_dev.py`.

Once this is up and running, we can send requests to the application and see the output.

```bash
curl -X PUT http://<cluster-ip-address>:8000/?name=Ray
# Hello, Ray! Hello, Ray!
```

For more complex runtime environments, you can specify the `runtime_env` via a YAML file; see [serve run](serve_cli.md#serve-run) for details.

:::{tip}

> If you need to upload local modules that reside in a directory that's not a subdirectory of your `working_dir`, use the `"py_modules"` field of `runtime_env`; e.g. `--runtime-env-json='{"working_dir": "/dir1", "py_modules": ["/dir2/my_module"]}'`.

Author: Yeah, should I delete it?

> If you need to upload local modules that reside in a directory that's not a subdirectory of your `working_dir`, use the `"py_modules"` field of `runtime_env`; e.g. `--runtime-env-json='{"working_dir": "/dir1", "py_modules": ["/dir2/my_module"]}'`.

Reviewer: I think so, or you could move it over there if it doesn't exist already.

> If you need to upload local modules that reside in a directory that's not a subdirectory of your `working_dir`, use the `"py_modules"` field of `runtime_env`; e.g. `--runtime-env-json='{"working_dir": "/dir1", "py_modules": ["/dir2/my_module"]}'`.

Author: Deleted it
